### PR TITLE
Extend Attachment-API

### PR DESCRIPTION
**Pull request by av-ps**
refs: https://github.com/metasfresh/metasfresh/issues/12356

### CODE REVIEW

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/v2/order/sales/SalesOrderRestController.java [hunk 1]
```diff
@@ -229,7 +229,7 @@ private JsonSalesOrderAttachment toSalesOrderAttachment(final int salesOrderId,
 		return JsonSalesOrderAttachment.builder()
 				.salesOrderId(String.valueOf(salesOrderId))
 				.id(AttachmentEntryId.getRepoId(entry.getId()))
-				.type(JsonAttachmentType.valueOf(entry.getType().toString()))
+				.type(JsonAttachmentSourceType.valueOf(entry.getType().name()))
```
Comment (metas-ts) [inline backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/v2/order/sales/SalesOrderRestController.java:232]: instead please add a little method with a switch, to make sure that renamings in one of the two enums won't break it
(also from de.metas.rest_api.v1.order.sales.SalesOrderRestController and de.metas.rest_api.v1.ordercandidates.impl.OrderCandidatesRestController)

**Review by metas-ts — CHANGES_REQUESTED** [review]

**Review by metas-ts — CHANGES_REQUESTED** [review]
i find the AttachmentEntryRepositoryTest very helpful!
Yet, AttachmentEntryRepository.retrieveAttachmentEntryDataResource is not covered and i still have my doubts if it would work with the new attachment-type :-)

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentEntryRepository.java [hunk 2]
```diff
@@ -128,6 +134,15 @@ private I_AD_AttachmentEntry retrieveAttachmentEntryRecordOutOfTrx(@NonNull fina
 	public byte[] retrieveAttachmentEntryData(@NonNull final AttachmentEntryId attachmentEntryId)
 	{
 		final I_AD_AttachmentEntry record = retrieveAttachmentEntryRecordInTrx(attachmentEntryId);
+		final String type = record.getType();
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentEntryRepository.java:137]: minor but important for future: pls convert to enum asap, i.e.
```
final AttachmentEntryType type = AttachmentEntryType.ofCode/ofNullableCode(record.getType());
```

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentEntryRepository.java [hunk 3]
```diff
@@ -291,4 +306,24 @@ public void delete(@NonNull final AttachmentEntry attachmentEntry)
...
+			if (filePath == null)
+			{
+				throw new AdempiereException("Cannot parse file path from url: " + url.toString());
+			}
+
+			return Files.readAllBytes(filePath);
+		}
+		catch (final IOException e)
+		{
+			throw new AdempiereException("Could not get binary data from url " + uri);
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/attachments/AttachmentEntryRepository.java:326]: don't shallow the `IOException e`. Use it as cause.

File: backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5621680_sys_gh12356_AD_AttachmentEntry_LocalFileURL_type.sql [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+;
+
+-- 2022-01-19T09:43:05.281Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+UPDATE AD_Column SET FieldLength=2,Updated=TO_TIMESTAMP('2022-01-19 11:43:05','YYYY-MM-DD HH24:MI:SS'),UpdatedBy=100 WHERE AD_Column_ID=557319
+;
+
+-- 2022-01-19T09:46:19.729Z
+-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
+INSERT INTO t_alter_column values('ad_attachmententry','Type','VARCHAR(2)',null,null)
```
Comment (teosarca) [inline backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5621680_sys_gh12356_AD_AttachmentEntry_LocalFileURL_type.sql:33]: loud thinking: shouldn't we make the Type mandatory?

Comment (teosarca) [reply backend/de.metas.adempiere.adempiere/migration/src/main/sql/postgresql/system/10-de.metas.adempiere/5621680_sys_gh12356_AD_AttachmentEntry_LocalFileURL_type.sql:33]: ahh, or it's already mandatory and here we just enlarge it to 2 chars.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/utils/JsonConverters.java [hunk 5]
```diff
@@ -69,4 +72,22 @@ public ExternalId fromJsonOrNull(@Nullable final JsonExternalId jsonExternalId)
 		}
 		return ExternalId.of(jsonExternalId.getValue());
 	}
+
+	@NonNull
+	public JsonAttachmentType toJsonAttachmentType(@NonNull final AttachmentEntryType type)
```
Comment (teosarca) [inline backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/utils/JsonConverters.java:77]: add a unit test that checks that we have an answer for each value of AttributeEntryType.values()

File: backend/de.metas.business.rest-api-impl/src/test/java/de/metas/rest_api/v1/ordercandidates/impl/OrderCandidatesRestController_misc_Test.java [hunk 6]
```diff
@@ -59,18 +60,18 @@ public static void afterAll()
...
-	 * Asserts that every {@link AttachmentEntry.Type} has a matching {@link JsonAttachmentType} and vice versa
+	 * Asserts that every {@link AttachmentEntryType} has a matching {@link JsonAttachmentType} and vice versa
 	 */
 	@Test
 	void jsonAttachmentTypes()
 	{
 		for (final JsonAttachmentType jsonAttachmentEntryType : JsonAttachmentType.values())
 		{
-			final AttachmentEntry.Type attachmentEntryType = AttachmentEntry.Type.valueOf(jsonAttachmentEntryType.toString());
+			final AttachmentEntryType attachmentEntryType = AttachmentEntryType.valueOf(jsonAttachmentEntryType.name());
```
Comment (teosarca) [inline backend/de.metas.business.rest-api-impl/src/test/java/de/metas/rest_api/v1/ordercandidates/impl/OrderCandidatesRestController_misc_Test.java:70]: if we need to convert from JSON to AttachmentEntryType in production code (not only tests) i would prefer to have a dedicated method which sits somewhere near the one which converts to JSON.

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/attachments/DocumentAttachmentsRestController.java [hunk 7]
```diff
@@ -183,50 +184,55 @@ private boolean isAllowDeletingAttachments()
...
+			default:
+				throw new AdempiereException("Invalid attachment entry")
+						.setParameter("reason", "invalid type")
+						.setParameter("type", type)
+						.setParameter("entry", entry);
 		}
 	}
 
-	private static ResponseEntity<byte[]> extractResponseEntryFromData(@NonNull final IDocumentAttachmentEntry entry)
+	private static ResponseEntity<StreamingResponseBody> extractResponseEntryFromData(@NonNull final IDocumentAttachmentEntry entry)
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/attachments/DocumentAttachmentsRestController.java:204]: do we need this method? don't we have it in helper?

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/attachments/DocumentAttachmentsRestController.java [hunk 8]
```diff
@@ -183,50 +184,55 @@ private boolean isAllowDeletingAttachments()
...
 		headers.set(HttpHeaders.CONTENT_DISPOSITION, "inline; filename=\"" + entryFilename + "\"");
 		headers.setCacheControl("must-revalidate, post-check=0, pre-check=0");
-		final ResponseEntity<byte[]> response = new ResponseEntity<>(entryData, headers, HttpStatus.OK);
-		return response;
+
+		return new ResponseEntity<>(entryData, headers, HttpStatus.OK);
 	}
 
-	private static ResponseEntity<byte[]> extractResponseEntryFromURL(@NonNull final IDocumentAttachmentEntry entry)
+	private static ResponseEntity<StreamingResponseBody> extractResponseEntryFromURL(@NonNull final IDocumentAttachmentEntry entry)
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/attachments/DocumentAttachmentsRestController.java:228]: do we need this method? don't we have it in helper?

File: misc/de-metas-common/de-metas-common-util/src/main/java/de/metas/common/util/FileUtil.java [hunk 9]
```diff
@@ -64,4 +72,57 @@ public String stripIllegalCharacters(@Nullable String filename)
...
+		if (path == null)
+		{
+			throw new RuntimeException("Couldn't parse path from:" + url);
+		}
+
+		return path;
+	}
+
+	@Nullable
+	private Path parseWindowsLocalPath(@NonNull final URL url)
```
Comment (teosarca) [inline misc/de-metas-common/de-metas-common-util/src/main/java/de/metas/common/util/FileUtil.java:96]: why Windows? What's specific to Windows?

btw, our production code runs on Linux.

File: misc/de-metas-common/de-metas-common-util/src/main/java/de/metas/common/util/FileUtil.java [hunk 10]
```diff
@@ -64,4 +72,57 @@ public String stripIllegalCharacters(@Nullable String filename)
...
+			return Paths.get(normalizedPath);
+		}
+		catch (final Throwable throwable)
+		{
+			return null;
+		}
+	}
+
+	@Nullable
+	private Path parseFileURL(@NonNull final URL url)
```
Comment (teosarca) [inline misc/de-metas-common/de-metas-common-util/src/main/java/de/metas/common/util/FileUtil.java:113]: what does this method do?
What's the difference between this one and parseWindowsLocalPath?
Pls make it obvious in the code (better method naming?)

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.business.rest-api-impl/src/test/java/de/metas/rest_api/v1/ordercandidates/impl/OrderCandidatesRestController_misc_Test.java (hunk 6 above)
Comment (cp-ps) [reply backend/de.metas.business.rest-api-impl/src/test/java/de/metas/rest_api/v1/ordercandidates/impl/OrderCandidatesRestController_misc_Test.java:70]: it's just the test AFAIS

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/attachments/DocumentAttachmentsRestController.java (hunk 7 above)
Comment (cp-ps) [reply backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/attachments/DocumentAttachmentsRestController.java:204]: yep.. leftovers :)

File: misc/de-metas-common/de-metas-common-util/src/main/java/de/metas/common/util/FileUtil.java (hunk 9 above)
Comment (cp-ps) [reply misc/de-metas-common/de-metas-common-util/src/main/java/de/metas/common/util/FileUtil.java:96]: yeah,,, but we run it on windows so it got in the way :) the "special" thing is the partition e.g. `C:`

**Review by metas-ts — APPROVED** [review]

**Pull request merged**